elastic/logstash
# Extended Flow Metrics

## Release notes

Extends the flow rates introduced to the Node Stats API in 8.5.0 (which included windows for `current` and `lifetime`) to include a Technology Preview of several additional windows such as `last_15_minutes`, `last_24_hours`, etc..

## What does this PR do?

1. breaks the initial `FlowMetric` implementation into an interface, a group of sharable components, and a net-unchanged concrete implementation `SimpleFlowMetric`.
2. ensures that flow metrics are rounded to a _precision_ instead of a _scale_ since we are looking for 3ish significant figures to make the rates meaningful and easily readable, not 3 decimal places as initially delivered (backport eligible)
3. introduce a new `ExtendedFlowMetric` that produces a number of policy-driven rates which are all marked as Technology Preview:
   > * `last_1_minute`: `1m`@`3s`
   > * `last_5_minutes`: `5m`@`15s`
   > * `last_15_minutes`: `15m`@`30s`
   > * `last_1_hour`: `1h` @ `60s`
   > * `last_24_hours`: `24h`@`15m`


## Why is it important/What is the impact to the user?

In 8.5.0 we introduced `current` and `lifetime` windows for all flow metrics, but these are made much more valuable when we have additional rates because they allow us to determine at a glance if things are actively improving or declining:

~~~ json
    "filter_throughput": {
      "current": 263.4,
      "last_1_minute": 32040,
      "last_5_minutes": 83270,
      "last_15_minutes": 97070,
      "last_1_hour": 91230,
      "last_24_hours": 92810,
      "lifetime": 90840
    },
~~~

## Checklist


## How to test this PR locally

The following pipeline will create extremely bursty throughput by injecting randomized sleeps during prime minutes, ensuring that the bursts are sustained for long enough that we can see them propagate through the rate windows.

~~~
input { generator { threads => 4 } }

filter {
  ruby {
    init => 'require "prime"'
    code => "
      if Prime.prime?(Time.now.min) && (Random.rand(100) <= 1)
        sleep(2.0 ** Random.rand(4.0))
      end
    "
  }
}

output { sink { } }
~~~

Once the pipeline is running, watch the Node Stats API noting that rate windows will not be present until sufficient time has elapsed from pipeline start for them to be meaningful (e.g., the `last_1_hour` will be present after the pipeline has been running for ~1 hour):

~~~
 watch "curl -XGET 'localhost:9600/_node/stats' | jq .flow"
~~~

Note that trace-level logging in `ExtendedFlowMetric` and `BaseFlowMetric` make it possible to reconstruct the retention and compaction of our windows.

## Related issues

- Closes #14570
